drivers/adc/aducm3029/aducm3029_adc.c

```diff
@@ -0,0 +1,178 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <drivers/adc/adi_adc.h>
+#include "aducm3029_adc.h"
+#include "error.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "util.h"
+
+#define ADUCM3029_DEV_ID 0
+
+struct adc_desc {
+	uint8_t		dev_mem[ADI_ADC_MEMORY_SIZE];
+	ADI_ADC_HANDLE	dev;
+	uint32_t	ch_mask;
+};
+
+int32_t aducm3029_adc_update_active_channels(struct adc_desc *desc,
+		uint32_t mask)
+{
+	if (!desc)
+		return -EINVAL;
+
+	desc->ch_mask = mask;
+
+	return SUCCESS;
+}
+
+int32_t aducm3029_adc_read(struct adc_desc *desc, uint16_t *buff,
+			   uint32_t nb_samples)
+{
+	ADI_ADC_BUFFER	adi_buff, *pbuff;
+	int32_t		ret;
+	uint32_t	i;
+
+	if (!desc)
+		return -EINVAL;
+
+	for (i = 0; i < nb_samples; i++) {
+		adi_buff.nChannels = desc->ch_mask;
+		adi_buff.pDataBuffer = buff + i * hweight8(desc->ch_mask);
+		adi_buff.nNumConversionPasses = 1;
+		adi_buff.nBuffSize = hweight8(desc->ch_mask) * sizeof(uint16_t);
+		ret = adi_adc_SubmitBuffer(desc->dev, &adi_buff);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+		ret = adi_adc_Enable(desc->dev, true);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+		ret = adi_adc_GetBuffer(desc->dev, &pbuff);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+		ret = adi_adc_Enable(desc->dev, false);
+		if (ret != ADI_ADC_SUCCESS)
+			return -ret;
+	}
+
+	return SUCCESS;
+}
+
+static int inline count_bits(uint32_t ch_mask)
+{
+	int count = 0;
+	while (ch_mask) {
+		count += ch_mask & 1;
+		ch_mask >>= 1;
+	}
+
+	return count;
+}
+
+/* Initialize the ADC */
+int32_t aducm3029_adc_init(struct adc_desc **desc, struct adc_init_param *param)
+{
+	UNUSED_PARAM(param);
+
+	ADI_ADC_RESULT	ret;
+	struct adc_desc	*ldesc;
+	bool		ready;
+
+	if (!desc)
+		return -EINVAL;
+
+	ldesc = (struct adc_desc *)calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		return -ENOMEM;
+
+	ret = adi_adc_Open(ADUCM3029_DEV_ID, ldesc->dev_mem,
+			   ADI_ADC_MEMORY_SIZE, &ldesc->dev);
+	if (ret != ADI_ADC_SUCCESS)
+		goto free_desc;
+
+	ret = adi_adc_PowerUp(ldesc->dev, true);
+	if (ret != ADI_ADC_SUCCESS)
+		goto close_adc;
+
+	ret = adi_adc_SetVrefSource(ldesc->dev, ADI_ADC_VREF_SRC_INT_2_50_V);
+	if (ret != ADI_ADC_SUCCESS)
+		goto close_adc;
```

Comment:
Maybe go to "adc_power_down"?

Reply:
added adi_adc_powerUP(ldesc->dev, false);
before closing the handler